File: palettier-ai/src/main/java/be/asmolabs/palettier/ai/ModelCatalog.java
package be.asmolabs.palettier.ai;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

/**
 * Dialogue avec le serveur Ollama local : ce qui est installe, et comment en installer
 * davantage.
 *
 * <p>Ollama publie tout cela sans authentification, sur la machine meme. Les autres
 * moteurs exposent des catalogues lies a un compte : pour eux, le nom du modele reste
 * une saisie libre.</p>
 */
@Component
public class ModelCatalog {

    private static final Logger log = LoggerFactory.getLogger(ModelCatalog.class);
    private static final Duration SHORT_TIMEOUT = Duration.ofSeconds(4);

    /** Un telechargement de modele se compte en gigaoctets : aucune limite de duree. */
    private static final Duration PULL_TIMEOUT = Duration.ofHours(2);

    private final String provider;
    private final String baseUrl;
    private final ObjectMapper json = new ObjectMapper();

    ModelCatalog(@Value("${spring.ai.model.chat:none}") String provider,
                 @Value("${spring.ai.ollama.base-url:http://localhost:11434}") String baseUrl) {
        this.provider = provider;
        this.baseUrl = baseUrl.replaceAll("/+$", "");
    }

    /**
     * Un modele installe.
     *
     * @param capabilities ce que le modele sait faire. La presence de {@code vision} est
     *                     decisive : sans elle, l'assistant ne peut pas lire vos photos.
     */
    public record ModelInfo(String name, long sizeBytes, List<String> capabilities) {

        public boolean supportsVision() {
            return capabilities.contains("vision");
        }

        public String sizeLabel() {
            return sizeBytes <= 0 ? "" : "%.1f Go".formatted(sizeBytes / 1e9);
        }
    }

    /** Avancement d'un telechargement. */
    public record PullProgress(String status, long completed, long total) {

        /** Fraction telechargee, ou -1 quand le serveur ne l'indique pas encore. */
        public double fraction() {
            return total > 0 ? (double) completed / total : -1;
        }
    }

    public String provider() {
        return provider;
    }

    public String baseUrl() {
        return baseUrl;
    }

    /** Vrai quand le moteur actif permet de lister et d'installer des modeles. */
    public boolean isManageable() {
        return "ollama".equalsIgnoreCase(provider);
    }

    /** Vrai si le serveur repond. */
    public boolean reachable() {
        return isManageable() && get("/api/tags") != null;
    }

    /** Modeles installes, avec leurs capacites. Liste vide si le serveur ne repond pas. */
    public List<ModelInfo> installed() {
        JsonNode tags = isManageable() ? get("/api/tags") : null;
        if (tags == null) {
            return List.of();
        }
        List<ModelInfo> models = new ArrayList<>();
        for (JsonNode model : tags.path("models")) {
            String name = model.path("name").asText("");
            if (!name.isBlank()) {
                models.add(new ModelInfo(name, model.path("size").asLong(0), capabilitiesOf(name)));
            }
        }
        models.sort((a, b) -> a.name().compareToIgnoreCase(b.name()));
        return List.copyOf(models);
    }

    /** Noms seuls, pour les listes deroulantes. */
    public List<String> available() {
        return installed().stream().map(ModelInfo::name).toList();
    }

    private List<String> capabilitiesOf(String model) {
        JsonNode details = post("/api/show", "{\"model\":\"" + model + "\"}");
        if (details == null) {
            return List.of();
        }
        List<String> capabilities = new ArrayList<>();
        details.path("capabilities").forEach(node -> capabilities.add(node.asText()));
        return List.copyOf(capabilities);
    }

    /**
     * Telecharge un modele et rend compte de l'avancement au fur et a mesure.
     *
     * <p>Ollama repond en flux : une ligne JSON par etape. On les transmet telles quelles
     * plutot que d'attendre la fin, sans quoi l'ecran resterait fige pendant plusieurs
     * gigaoctets.</p>
     *
     * @param onProgress recoit chaque etape ; appele depuis le fil qui invoque la methode
     * @return vrai si le telechargement s'est acheve
     */
    public boolean pull(String model, Consumer<PullProgress> onProgress) {
        if (!isManageable() || model == null || model.isBlank()) {
            return false;
        }
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/api/pull"))
                .timeout(PULL_TIMEOUT)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(
                        "{\"model\":\"" + model.trim() + "\",\"stream\":true}", StandardCharsets.UTF_8))
                .build();

        try (HttpClient client = HttpClient.newBuilder().connectTimeout(SHORT_TIMEOUT).build()) {
            HttpResponse<java.io.InputStream> response =
                    client.send(request, HttpResponse.BodyHandlers.ofInputStream());
            if (response.statusCode() != 200) {
                onProgress.accept(new PullProgress("refuse par le serveur (" + response.statusCode() + ")", 0, 0));
                return false;
            }

            boolean success = false;
            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(response.body(), StandardCharsets.UTF_8))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (line.isBlank()) {
                        continue;
                    }
                    JsonNode step = json.readTree(line);
                    if (step.hasNonNull("error")) {
                        onProgress.accept(new PullProgress(step.get("error").asText(), 0, 0));
                        return false;
                    }
                    String status = step.path("status").asText("");
                    onProgress.accept(new PullProgress(status,
                            step.path("completed").asLong(0), step.path("total").asLong(0)));
                    success = "success".equals(status);
                }
            }
            return success;
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return false;
        } catch (Exception e) {
            log.debug("Telechargement du modele {} impossible", model, e);
            onProgress.accept(new PullProgress("echec : " + e.getMessage(), 0, 0));
            return false;
        }
    }

    // --- Acces HTTP --------------------------------------------------------

    private JsonNode get(String path) {
        return call(HttpRequest.newBuilder().uri(URI.create(baseUrl + path)).GET());
    }

    private JsonNode post(String path, String body) {
        return call(HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + path))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(body, StandardCharsets.UTF_8)));
    }

    private JsonNode call(HttpRequest.Builder builder) {
        try (HttpClient client = HttpClient.newBuilder().connectTimeout(SHORT_TIMEOUT).build()) {
            HttpResponse<String> response = client.send(
                    builder.timeout(SHORT_TIMEOUT).build(), HttpResponse.BodyHandlers.ofString());
            return response.statusCode() == 200 ? json.readTree(response.body()) : null;
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return null;
        } catch (Exception e) {
            log.debug("Serveur Ollama injoignable sur {}", baseUrl, e);
            return null;
        }
    }
}

File: palettier-kmp/ai/src/commonMain/kotlin/be/asmolabs/palettier/ai/ModelCatalog.kt
package be.asmolabs.palettier.ai

import io.ktor.client.HttpClient
import io.ktor.client.request.get
import io.ktor.client.request.post
import io.ktor.client.request.preparePost
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.client.statement.bodyAsChannel
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import io.ktor.utils.io.readUTF8Line
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long
import kotlinx.serialization.json.put

/**
 * Un modele installe.
 *
 * @param capabilities ce que le modele sait faire. La presence de "vision" est decisive :
 *                     sans elle, l'assistant ne peut pas lire vos photos.
 */
data class ModelInfo(val name: String, val sizeBytes: Long, val capabilities: List<String>) {

    val supportsVision: Boolean get() = "vision" in capabilities

    val sizeLabel: String get() = if (sizeBytes <= 0) "" else {
        val go = sizeBytes / 1e9
        "${(go * 10).toLong() / 10.0} Go"
    }
}

/** Avancement d'un telechargement. */
data class PullProgress(val status: String, val completed: Long = 0, val total: Long = 0) {

    /** Fraction telechargee, ou -1 quand le serveur ne l'indique pas encore. */
    val fraction: Double get() = if (total > 0) completed.toDouble() / total else -1.0

    val done: Boolean get() = status == "success"
}

/**
 * Dialogue avec le serveur Ollama local : ce qui est installe, et comment en installer
 * davantage.
 *
 * <p>Ollama publie tout cela sans authentification, sur la machine meme. Les autres
 * moteurs exposent des catalogues lies a un compte : pour eux, le nom du modele reste une
 * saisie libre, et ce catalogue-ci se declare simplement injoignable.</p>
 */
class ModelCatalog(
    private val client: HttpClient,
    private val store: AiSettingsStore,
) {

    private val json = Json { ignoreUnknownKeys = true }

    /** Vrai quand le moteur actif permet de lister et d'installer des modeles. */
    suspend fun isManageable(): Boolean = store.current().provider == AiProvider.OLLAMA

    private suspend fun baseUrl(): String =
        store.current().baseUrl.ifBlank { OllamaEngine.DEFAULT_URL }.trimEnd('/')

    /** Vrai si le serveur repond. */
    suspend fun reachable(): Boolean = isManageable() && get("/api/tags") != null

    /** Modeles installes, avec leurs capacites. Liste vide si le serveur ne repond pas. */
    suspend fun installed(): List<ModelInfo> {
        val tags = if (isManageable()) get("/api/tags") else null
        val models = tags?.jsonObject?.get("models")?.jsonArray ?: return emptyList()

        return models.mapNotNull { entry ->
            val name = entry.jsonObject["name"]?.jsonPrimitive?.content.orEmpty()
            if (name.isBlank()) null else ModelInfo(
                name = name,
                sizeBytes = entry.jsonObject["size"]?.jsonPrimitive?.long ?: 0L,
                capabilities = capabilitiesOf(name),
            )
        }.sortedBy { it.name.lowercase() }
    }

    private suspend fun capabilitiesOf(model: String): List<String> {
        val details = post("/api/show", buildJsonObject { put("model", model) }.toString())
        return details?.jsonObject?.get("capabilities")?.jsonArray
            ?.map { it.jsonPrimitive.content }
            ?: emptyList()
    }

    /**
     * Telecharge un modele et rend compte de l'avancement au fur et a mesure.
     *
     * <p>Ollama repond en flux : une ligne JSON par etape. On les transmet telles quelles
     * plutot que d'attendre la fin, sans quoi l'ecran resterait fige pendant plusieurs
     * gigaoctets.</p>
     */
    fun pull(model: String): Flow<PullProgress> = flow {
        if (!isManageable() || model.isBlank()) {
            emit(PullProgress("Le telechargement n'est possible qu'avec Ollama."))
            return@flow
        }

        val body = buildJsonObject {
            put("model", model.trim())
            put("stream", true)
        }.toString()

        client.preparePost("${baseUrl()}/api/pull") {
            contentType(ContentType.Application.Json)
            setBody(body)
        }.execute { response ->
            if (!response.status.isSuccess()) {
                emit(PullProgress("refuse par le serveur (${response.status.value})"))
                return@execute
            }
            val channel = response.bodyAsChannel()
            while (true) {
                val line = channel.readUTF8Line() ?: break
                if (line.isBlank()) continue
                val step = runCatching { json.parseToJsonElement(line).jsonObject }.getOrNull() ?: continue

                step["error"]?.jsonPrimitive?.content?.let {
                    emit(PullProgress(it))
                    return@execute
                }
                emit(
                    PullProgress(
                        status = step["status"]?.jsonPrimitive?.content.orEmpty(),
                        completed = step["completed"]?.jsonPrimitive?.long ?: 0L,
                        total = step["total"]?.jsonPrimitive?.long ?: 0L,
                    )
                )
            }
        }
    }

    // --- Acces HTTP --------------------------------------------------------

    private suspend fun get(path: String): JsonElement? = read { client.get("${baseUrl()}$path") }

    private suspend fun post(path: String, body: String): JsonElement? = read {
        client.post("${baseUrl()}$path") {
            contentType(ContentType.Application.Json)
            setBody(body)
        }
    }

    /** Un serveur eteint n'est pas un incident : c'est le cas normal avant installation. */
    private suspend fun read(call: suspend () -> HttpResponse): JsonElement? = runCatching {
        val response = call()
        if (response.status.isSuccess()) json.parseToJsonElement(response.bodyAsText()) else null
    }.getOrNull()
}
